### DISCUSSION TITLE
Configuration file errors when using `cypress run`, no issue with `cypress open`

### DISCUSSION THREAD

**Discussion by jhulme** [cypress-io/cypress#25770] [Questions and Help]
I have been adding functionality to one of our projects, in order to read in secrets stored in a Hashicorp vault instance. I am taking advantage of the [node-vault](https://www.npmjs.com/package/node-vault) package. We have a number of environment specific configuration files, each retrieving the appropriate values from vault. For example a rough outline of the file `cypress/config/cypress.qa.config.js` is below

```
const { defineConfig } = require("cypress");
const defaultConfig = require("../../cypress.config")
const vault = require('node-vault')

var vaultdata
// Set URL for vault
var vault_options = {
  endpoint: "https://qa.internalvault.com"
}

v = vault(vault_options)

v.approleLogin({ role_id: '1234', secret_id: 5678'}).then(() => {
  console.log('connecting to vault...')
  v.read('secret/organisation/cypress-creds/qa').then((result) => {
    vaultdata = result.data
  })
})

module.exports = defineConfig({
    ...defaultConfig,
    env: {
        alternativeTestUI: "https://testurl-ui.com",
    },
    e2e: {
        baseUrl: 'https://testurl.com',
        setupNodeEvents(on, config) {
          config.env.username = vaultdata.username
          config.env.password = vaultdata.password

          return config
        }
    }
})
```

The issue appears to be that when I launch using `cypress open --config-file cypress/config/cypress.qa.config.js` the data from vault is extracted and the code within `setupNodeEvents()` sets the values of `config.env.username` and `config.env.password` successfully.

I can then see that these values have been set correctly when viewing the Resolved Configuration within the Project Settings.

However if I repeat the process using the headless `cypress run --config-file cypress/config/cypress.qa.config.js` I receive the following error:

```
TypeError: Cannot read properties of undefined (reading 'username')
    at setupNodeEvents (/home/user/demoproject/cypress/config/cypress.qa.config.js:109:53)
```

Which appears to indicate that `vaultdata.username` has no value set. 

I assume there is something wrong with how I have structured my code, but I would appreciate any advice on resolving the issue as I have hit a bit of a blocker here and I can't seem to find a method of resolving the problem. Checking the documentation indicates that the `setupNodeEvents()` function can return a Promise rather than just a synchronous value, but I have been unable to find a working approach using this structure so far.

**Comment by jhulme** [comment]
I've had some time to take another look into this and it looks likely I am falling foul of an existing bug. 

I have adjusted my code to move the communications with vault into the `(on before:run ` event, however it would appear that this functionality is currently broken at present under the following items:

* https://github.com/cypress-io/cypress/issues/24403
* https://github.com/cypress-io/cypress/issues/22360

For now I will have to return to managing my credentials via environment variables rather than pulling them from a vault instance until this is resolved.

**Reply by HOPELESSNESSLYR** [reply]
TypeError: Cannot read properties of undefined (reading 'username') I have a same problem. how to solve it ?Thank you!
